### ISSUE TITLE
Replace separate math/operator nodes with generics

### ISSUE THREAD

**Issue by Frozenreflex** [Neos-Metaverse/NeosPublic#3615]
### Is your feature request related to a problem? Please describe.

Most math nodes (such as add, mul, div, multiadd, cast, etc) are implemented separately for each type, rather than using a generic. Ingame, this mostly affects adding components, by flooding the component attacher window with all of the different variations if a section with said variations is opened.

### Relevant issues

None that I could find when searching "generic"

### Describe the solution you'd like

Replace such nodes with generics, similar to how individual type registers were replaced with ValueRegister<T>

### Describe alternatives you've considered

None available

**Comment by Earthmark** [comment]
This would require dotnet 6, otherwise there isn't a type to generalize the math operations over in terms of a generic constraint. See https://devblogs.microsoft.com/dotnet/preview-features-in-net-6-generic-math/

One could manually be added but that's a lot of pain and wrappers for very little gain. In practice the LogiX tool should handle most of these conversations for you.

What are you doing such that the LogiX tip's auto overrides are not working?

**Comment by Frozenreflex** [comment]
I'm not doing anything related to Logix, I'm decompiling the game and trying to fix the frequent crashes, and seeing huge lists of different variations of math nodes is annoying, and in general something like this should be done with generics, like how ValueRegister is. However, I have had times where I've needed to use the component attacher (don't remember why, I think I was trying to cast something) to attach logix nodes, and it was a pain to scroll through.

**Comment by Earthmark** [comment]
I recall in a conversation the engine is code gen heavy, it's possible the source uses templating for those but I don't know. Generics over those would be interesting for sure, I do think that's a dotnet 6 possibility but I'm doubtful for the now case.

Do you have GitHub issues for the crashes you are experiencing? That way the fix gets put into the core build, instead of your own.

**Comment by Frozenreflex** [comment]
A few of the logs in #3265 have the same issue I have, engine unresponsive message where the WorldSessionState is stuck at WaitingForSyncThreadEvent, and I've posted my findings up to this point there.

**Comment by ProbablePrime** [comment]
Feature requests are always welcome, however sometimes feature requests are actually bug reports in disguise. 

Respectfully, I feel this might be one of those cases.

> this mostly affects adding components, by flooding the component attacher window with all of the different variations if a section with said variations is opened.

Ideally you shouldn't need to add Logix nodes using the component attacher, However if you should need to do that, then we can make that better with a new component attacher which will have better handling. Additionally the new component attacher(part of the UI Re-work) will result in some reclassification of Components. So things like Uncat or Logix being cluttered may be resolved. 

Due to both of these items, I'm not sure the component attacher being a long list here is a problem. Its something we have fixes in mind for.

> fix the frequent crashes
Make sure you have issues open for **all** of these. I've seen some from you. Its up to us to fix these mostly. 

> decompiling the game and trying to fix the frequent crashes, and seeing huge lists of different variations of math nodes is annoying
The problem here is that you're decompiling it, this isn't an intended flow as Earthmark says, we use a ton of code generation for those so in the actual source code its no bother. I'm sorry that you're finding decompiling the game is annoying but generically I don't recommend doing this as it doesn't give you a complete picture of the situation. 

I feel this problem will go away if we fix your frequent crashes. 


> I have had times where I've needed to use the component attacher (don't remember why, I think I was trying to cast something) to attach logix nodes, and it was a pain to scroll through.

If you can remember these occurrences let us know, this might be a case where we can improve our overloading(when nodes swap between types) or provide a better experience. Remember that the Logix Node menu is also due for a rework. 

In summary, a lot of the problems you're experiencing shouldn't need to occur and if they do its a sign of either a bug or a workflow that we can improve with other features. 

In these cases, I often link the XYProblem website which is https://xyproblem.info. This has sometimes been seen as dismissive or rude but that's not my intention. Its intended to further the discussion.  Applying it to this issue:
- Y is replacing our math nodes with generics
- X is:
    - Frequent Crashes
    - The state of the Component attacher
    - The state of the Logix node Selector.

I think fixing all the items in the X list is a much better course of action.

**Comment by Frozenreflex** [comment]
I want to point out that I never mentioned the crashing issue in my initial issue, I only mentioned it after I was asked. This issue should not have anything to do with my crashing. However, now that I've had *a* response to my crashing issue for the first time in three months or so, please message me on Discord at Fro Zen#8796 so that I can finally get the crashes fixed.
> Its up to us to fix these mostly.

I'd take this a bit more seriously if the last full-fledged update wasn't pushed three months ago, and if I actually got communication when I asked for help, which is why I feel the need to decompile the game to fix it myself.
> This has sometimes been seen as dismissive or rude but that's not my intention

This absolutely feels dismissive.

**Comment by ProbablePrime** [comment]
> please message me on Discord at Fro Zen#8796 so that I can finally get the crashes fixed.

please open a GH issue with logs for each of your crashes. We need GH to act.

> wasn't pushed three months ago

You can read announcements for why that's happening.

> This absolutely feels dismissive.

As I said, this isn't my intention, its still an idea. I'm just unclear on why we'd need to do this.

**Comment by Frozenreflex** [comment]
> please open a GH issue with logs for each of your crashes. We need GH issues to act. Discord messages unfortunately don't cut it as we need a permanent record of each issue on the issue tracker.

I've posted plenty myself already, and I believe I've already shown why said logs also aren't very helpful (#3525).

**Comment by ProbablePrime** [comment]
That's not a log for an issue, that's again an implementation detail for Neos and our developers. Please open issues for crashes you're experiencing. If they're already open that's great. Thank you.

**Comment by Frozenreflex** [comment]
> That's not a log for an issue

Rephrase: Already posted logs of the issue elsewhere, but those logs are pretty meaningless because of #3525.
Here's everything I could find, including from others.
#3412
#3507
From #3265:
https://github.com/Neos-Metaverse/NeosPublic/files/7761746/q3.-.2021.11.10.1253.-.2021-12-22.03_15_48.log.txt
https://github.com/Neos-Metaverse/NeosPublic/files/7671915/FOXSTATION.-.2021.11.10.1265.-.2021-12-07.16_24_52.log
https://github.com/Neos-Metaverse/NeosPublic/files/7505259/DESKTOP-E9C0718_-_2021.11.7.794_-_2021-11-08_17_30_34.log
https://github.com/Neos-Metaverse/NeosPublic/files/7505282/DESKTOP-E9C0718_-_2021.11.7.794_-_2021-11-08_18_11_52.log
https://github.com/Neos-Metaverse/NeosPublic/files/7505284/DESKTOP-E9C0718_-_2021.11.7.794_-_2021-11-08_18_24_21.log
https://github.com/Neos-Metaverse/NeosPublic/files/7505285/DESKTOP-E9C0718_-_2021.11.7.794_-_2021-11-08_18_38_35.log
https://github.com/Neos-Metaverse/NeosPublic/files/7502129/RYUVI-TOWER.-.2021.11.7.784.-.2021-11-08.20_54_52.log
https://github.com/Neos-Metaverse/NeosPublic/files/7493159/lockup1.log
In addition, here's a dump of some of my log folder sorted using ``grep -ir "main thread stacktrace"``, but since all of them have mods enabled, including my own debugging statements, you won't accept them per the mod and plugin policy, and again, even if you did accept them, they're pretty useless anyway because of the issue I mentioned before.
[FROZENREFLEX - 2021.11.10.1253 - 2022-01-11 10_25_50.log](https://github.com/Neos-Metaverse/NeosPublic/files/8181986/FROZENREFLEX.-.2021.11.10.1253.-.2022-01-11.10_25_50.log)
[FROZENREFLEX - 2022.1.28.1335 - 2022-01-31 16_01_37.log](https://github.com/Neos-Metaverse/NeosPublic/files/8182003/FROZENREFLEX.-.2022.1.28.1335.-.2022-01-31.16_01_37.log)
[FROZENREFLEX - 2021.11.10.1253 - 2022-01-25 21_58_10.log](https://github.com/Neos-Metaverse/NeosPublic/files/8182014/FROZENREFLEX.-.2021.11.10.1253.-.2022-01-25.21_58_10.log)
[FROZENREFLEX - 2022.1.28.1335 - 2022-02-08 19_29_42.log](https://github.com/Neos-Metaverse/NeosPublic/files/8182025/FROZENREFLEX.-.2022.1.28.1335.-.2022-02-08.19_29_42.log)
[FROZENREFLEX - 2021.11.10.1253 - 2022-01-26 22_11_05.log](https://github.com/Neos-Metaverse/NeosPublic/files/8182027/FROZENREFLEX.-.2021.11.10.1253.-.2022-01-26.22_11_05.log)
[FROZENREFLEX - 2021.11.10.1253 - 2022-01-11 07_37_25.log](https://github.com/Neos-Metaverse/NeosPublic/files/8182034/FROZENREFLEX.-.2021.11.10.1253.-.2022-01-11.07_37_25.log)
[FROZENREFLEX - 2022.1.28.1335 - 2022-01-28 22_07_53.log](https://github.com/Neos-Metaverse/NeosPublic/files/8182060/FROZENREFLEX.-.2022.1.28.1335.-.2022-01-28.22_07_53.log)
[FROZENREFLEX - 2022.1.28.1335 - 2022-02-08 22_33_19.log](https://github.com/Neos-Metaverse/NeosPublic/files/8182076/FROZENREFLEX.-.2022.1.28.1335.-.2022-02-08.22_33_19.log)
[FROZENREFLEX - 2021.11.10.1253 - 2022-01-10 18_06_32.log](https://github.com/Neos-Metaverse/NeosPublic/files/8182086/FROZENREFLEX.-.2021.11.10.1253.-.2022-01-10.18_06_32.log)
[FROZENREFLEX - 2021.11.10.1253 - 2022-01-11 07_16_09.log](https://github.com/Neos-Metaverse/NeosPublic/files/8182094/FROZENREFLEX.-.2021.11.10.1253.-.2022-01-11.07_16_09.log)
[FROZENREFLEX - 2022.1.28.1335 - 2022-01-31 15_37_30.log](https://github.com/Neos-Metaverse/NeosPublic/files/8182098/FROZENREFLEX.-.2022.1.28.1335.-.2022-01-31.15_37_30.log)
[FROZENREFLEX - 2021.11.10.1253 - 2022-01-11 08_23_56.log](https://github.com/Neos-Metaverse/NeosPublic/files/8182101/FROZENREFLEX.-.2021.11.10.1253.-.2022-01-11.08_23_56.log)
I've been digging at this for over three months now, and so far Github issues have been fairly useless, it's been three months of absolutely nothing, and I've posted and seen others post this issue, which is why I said contact me through Discord so that we could work on actually fixing it, and to not flood this issue with unrelated stuff. I am fully aware of whats going on between y'all and Karel, and I don't care, I would prefer to not crash every five minutes regardless of if Karel is throwing a fit, so whether it's through a mod or the first actual update in three months that isn't an RCE fix or a single character change in a string, I want a fix sooner rather than later, and I don't have confidence that it would be sooner if I left it only at Github issues.

**Comment by ProbablePrime** [comment]
Thanks for linking your issues. Make sure you have github issues open for each issue and then the Neos team will take a look at the logs and deem the usefulness of the logs from there. We'll likely ask for additional information or questions if we need them. 

Thank you.
